services/dynamic-sidecar/tests/unit/test_modules_outputs_event_filter.py

```diff
@@ -0,0 +1,235 @@
+) -> Iterator[AsyncMock]:
+    async def _mock_upload_outputs(*args, **kwargs) -> None:
+        pass
+
+    yield mocker.patch.object(
+        outputs_manager, "port_key_content_changed", side_effect=_mock_upload_outputs
+    )
+
+
+@pytest.fixture
+def mock_delay_policy() -> BaseDelayPolicy:
+    VERY_FAST_BASE = 0.001
+    SLOW_LONG_INTERVAL = VERY_FAST_BASE * 100
+
+    # pylint:disable=no-self-use
+    class FastPolicy(BaseDelayPolicy):
+        def get_min_interval(self) -> NonNegativeFloat:
+            return VERY_FAST_BASE
+
+        def get_wait_interval(self, dir_size: NonNegativeInt) -> NonNegativeFloat:
+            return VERY_FAST_BASE if dir_size == 0 else SLOW_LONG_INTERVAL
+
+    return FastPolicy()
+
+
+@pytest.fixture
+def mock_get_dir_size(mocker: MockerFixture) -> Iterator[AsyncMock]:
+    yield mocker.patch(
+        "simcore_service_dynamic_sidecar.modules.outputs._event_filter.get_dir_size",
+        return_value=1,
+    )
+
+
+@pytest.fixture
+async def event_filter(
+    outputs_manager: OutputsManager, mock_delay_policy: BaseDelayPolicy
+) -> AsyncIterator[EventFilter]:
+    event_filter = EventFilter(
+        outputs_manager=outputs_manager, delay_policy=mock_delay_policy
+    )
+    await event_filter.start()
+    yield event_filter
+    await event_filter.shutdown()
+
+
+# UTILS
+
+
+async def _wait_for_event_to_trigger(event_filter: EventFilter) -> None:
+    await asyncio.sleep(event_filter.delay_policy.get_min_interval() * 2)
+
+
+async def _wait_for_event_to_trigger_big_directory(event_filter: EventFilter) -> None:
+    await asyncio.sleep(event_filter.delay_policy.get_wait_interval(1) * 2)
+
+
+# TESTS
+
+
+async def test_event_triggers_once(
+    event_filter: EventFilter,
+    port_key_1: str,
+    mocked_port_key_content_changed: AsyncMock,
+):
+    # event triggers once
+    await event_filter.enqueue(port_key_1)
+    await _wait_for_event_to_trigger(event_filter)
+    assert mocked_port_key_content_changed.call_count == 1
+
+    await _wait_for_event_to_trigger(event_filter)
+
+    # event triggers a second time
+    await event_filter.enqueue(port_key_1)
+    await _wait_for_event_to_trigger(event_filter)
+    assert mocked_port_key_content_changed.call_count == 2
+
+
+async def test_trigger_once_after_event_chain(
+    event_filter: EventFilter,
+    port_key_1: str,
+    mocked_port_key_content_changed: AsyncMock,
+):
+    for _ in range(100):
+        await event_filter.enqueue(port_key_1)
+    await _wait_for_event_to_trigger(event_filter)
+    assert mocked_port_key_content_changed.call_count == 1
+
+
+async def test_always_trigger_after_delay(
+    mock_get_dir_size: AsyncMock,
+    event_filter: EventFilter,
+    port_key_1: str,
+    mocked_port_key_content_changed: AsyncMock,
+):
+    # event triggers once
+    await event_filter.enqueue(port_key_1)
+    await _wait_for_event_to_trigger(event_filter)
+    assert mocked_port_key_content_changed.call_count == 0
```

Comment:
why here is not ``call_count==1``? I am missing something :-(

Reply:
Nice find, I've corrected the test, it makes sense.